File: banquet-pilot/src/core/solver.js
/** Static full-seat enumerator (no UI). */

import { seatsForLayout } from "./geometry.js";
import { isWin } from "./rules.js";

/**
 * @param {string[]} arr
 * @returns {Generator<string[]>}
 */
function* permutations(arr) {
  const a = arr.slice();
  const n = a.length;
  if (n === 0) {
    yield [];
    return;
  }
  const c = new Array(n).fill(0);
  yield a.slice();
  let i = 0;
  while (i < n) {
    if (c[i] < i) {
      const k = i % 2 === 0 ? 0 : c[i];
      const tmp = a[k];
      a[k] = a[i];
      a[i] = tmp;
      yield a.slice();
      c[i] += 1;
      i = 0;
    } else {
      c[i] = 0;
      i += 1;
    }
  }
}

/**
 * @param {string[]} characters
 * @param {string[]} seats
 * @returns {Generator<Record<string, string>>}
 */
function* fullAssignments(characters, seats) {
  if (characters.length !== seats.length) {
    throw new Error(
      `enumerator requires |characters|==|seats|, got ${characters.length} vs ${seats.length}`,
    );
  }
  for (const perm of permutations(seats)) {
    /** @type {Record<string, string>} */
    const asg = {};
    for (let i = 0; i < characters.length; i++) asg[characters[i]] = perm[i];
    yield asg;
  }
}

/**
 * @param {object} level
 * @param {{ calm?: Iterable<string>|null }} opts
 */
export function enumerateSolutions(level, { calm = null } = {}) {
  const characters = level.characters;
  const seats = level.seats || seatsForLayout(level.layout);
  /** @type {Record<string, string>[]} */
  const solutions = [];
  for (const asg of fullAssignments(characters, seats)) {
    if (isWin(level, asg, { calm })) solutions.push(asg);
  }
  return solutions;
}

/**
 * @param {object} level
 * @param {{ calm?: Iterable<string>|null }} opts
 * @returns {[number, number]}
 */
export function countSolutions(level, { calm = null } = {}) {
  const characters = level.characters;
  const seats = level.seats || seatsForLayout(level.layout);
  let n = 0;
  let total = 0;
  for (const asg of fullAssignments(characters, seats)) {
    total += 1;
    if (isWin(level, asg, { calm })) n += 1;
  }
  return [n, total];
}

File: banquet-pilot/src/core/solver.py
"""Static full-seat enumerator (no UI)."""

from __future__ import annotations

import itertools
from typing import Any, Dict, Iterator, List, Mapping, Optional, Sequence, Set, Tuple

from .geometry import seats_for_layout
from .rules import is_win


def _full_assignments(
    characters: Sequence[str], seats: Sequence[str]
) -> Iterator[Dict[str, str]]:
    if len(characters) != len(seats):
        raise ValueError(
            f"enumerator requires |characters|==|seats|, got {len(characters)} vs {len(seats)}"
        )
    for perm in itertools.permutations(seats):
        yield {c: s for c, s in zip(characters, perm)}


def enumerate_solutions(
    level: Mapping[str, Any],
    *,
    calm: Optional[Set[str]] = None,
) -> List[Dict[str, str]]:
    characters = list(level["characters"])
    seats = list(level.get("seats") or seats_for_layout(level["layout"]))
    solutions: List[Dict[str, str]] = []
    for asg in _full_assignments(characters, seats):
        if is_win(level, asg, calm=calm):
            solutions.append(asg)
    return solutions


def count_solutions(
    level: Mapping[str, Any],
    *,
    calm: Optional[Set[str]] = None,
) -> Tuple[int, int]:
    """Return (solution_count, permutations_checked)."""
    characters = list(level["characters"])
    seats = list(level.get("seats") or seats_for_layout(level["layout"]))
    n = 0
    total = 0
    for asg in _full_assignments(characters, seats):
        total += 1
        if is_win(level, asg, calm=calm):
            n += 1
    return n, total
